crates/sui-core/src/authority/authority_store.rs

```diff
@@ -785,15 +785,42 @@ impl<S: Eq + Debug + Serialize + for<'de> Deserialize<'de>> SuiDataStore<S> {
         let transaction_digest = certificate.digest();
         let mut temporary_store =
             TemporaryStore::new(Arc::new(&self), input_objects, *transaction_digest);
-        for (_, (object, kind)) in mutated_objects {
+
+        // TemporaryStore::into_inner will increment object versions, and requires that this
+        // operation is a true increment. Roll back modified objects' versions so that this remains
+        // the case.
+        let prev_versions: HashMap<ObjectID, SequenceNumber> = effects
+            .effects
+            .modified_at_versions
```

Comment:
Ah I think I see whats up (from my confusion in the other comment). Do we need this `modified_at_versions` because of dynamic child objects?

Reply:
@amnn we should be able to delete the remaining gateway code pretty soon fyi

Reply:
Happy days!  Yeah, I figured that would be the case, note that `modified_at_versions` will still be necessary for `revert_state_update` to handle dynamic child access as @tnowacki mentioned.